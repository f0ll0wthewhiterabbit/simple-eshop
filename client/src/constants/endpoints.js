export const PRODUCTS_API = '/data/products.json'
export const USERS_API = '/data/users.json'
