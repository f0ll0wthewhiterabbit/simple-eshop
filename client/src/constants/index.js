export * from './pagePaths'
export * from './endpoints'
export * from './fields'
export * from './roles'
export * from './pageLimits'
