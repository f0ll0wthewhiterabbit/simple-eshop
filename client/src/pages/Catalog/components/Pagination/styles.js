import styled from 'styled-components'
import { IconButton } from '@material-ui/core'

export const Wrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: baseline;
  flex-wrap: nowrap;
`

export const PaginationList = styled.ul`
  padding: 0;
  list-style: none;
  display: flex;
  justify-content: center;
  margin: 0;
`

export const PaginationItem = styled.li`
  margin-right: 18px;

  &:last-child {
    margin-right: 0;
  }
`

export const PaginationButton = styled(IconButton)`
  border-radius: 0;
  min-width: 20px;
  height: 28px;
  padding: 0;
  background-color: transparent;
  font-family: 'Montserrat', sans-serif;
  color: #838383;
  font-size: 12px;
  font-weight: 500;
  position: relative;

  &::after {
    display: block;
    position: absolute;
    left: 5px;
    bottom: -5px;
    width: 10px;
    height: 2px;
    background: #000;
    content: '';
    opacity: 0;
    transition: all 0.3s;
  }

  ${props =>
    props.active &&
    `
    color: #000;

    &::after {
      opacity: 1;
    }
  `}

  &:hover {
    background-color: transparent;
    color: #000;

    &::after {
      opacity: 1;
    }

    ${props =>
      props.active &&
      `
      cursor: default;
    `}
  }

  &:first-child {
    svg {
      width: 0.8em;
      height: 0.8em;
    }
  }

  &:last-child {
    svg {
      width: 0.8em;
      height: 0.8em;
    }
  }
`

export const Label = styled.span`
  font-family: 'Montserrat', sans-serif;
  color: #838383;
  font-size: 12px;
  font-weight: 500;
  margin-right: 24px;
`
