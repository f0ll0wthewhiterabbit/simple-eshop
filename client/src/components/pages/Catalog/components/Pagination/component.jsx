import React from 'react'
import PropTypes from 'prop-types'
import ChevronRightIcon from '@material-ui/icons/ChevronRight'
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft'

import { PaginationList, PaginationItem, PaginationButton } from './styles'

const Pagination = ({ currentPage, totalPages, filter, fetchProducts }) => {
  const pageNumbers = Array.from({ length: totalPages }, (v, k) => k + 1)

  const handlePageNumberClick = evt => {
    fetchProducts(Number(evt.target.innerText), null, filter)
  }

  const handlePreviousPageClick = () => {
    fetchProducts(currentPage - 1, null, filter)
  }
  const handleNextPageClick = () => {
    fetchProducts(currentPage + 1, null, filter)
  }

  const isPageNumberInPagination = pageNumber => {
    return (
      pageNumber === 1 ||
      pageNumber === totalPages ||
      (pageNumber >= currentPage - 2 && pageNumber <= currentPage + 2)
    )
  }

  return (
    <PaginationList>
      <PaginationItem>
        <PaginationButton
          aria-label="delete"
          size="small"
          disabled={currentPage === 1 ? true : undefined}
          onClick={handlePreviousPageClick}
        >
          <ChevronLeftIcon />
        </PaginationButton>
      </PaginationItem>
      {pageNumbers.map(
        pageNumber =>
          isPageNumberInPagination(pageNumber) && (
            <PaginationItem key={pageNumber}>
              <PaginationButton
                type="button"
                active={pageNumber === currentPage ? 1 : undefined}
                onClick={pageNumber === currentPage ? undefined : handlePageNumberClick}
              >
                {pageNumber}
              </PaginationButton>
            </PaginationItem>
          )
      )}
      <PaginationItem>
        <PaginationButton
          aria-label="delete"
          size="small"
          disabled={currentPage === totalPages ? true : undefined}
          onClick={handleNextPageClick}
        >
          <ChevronRightIcon />
        </PaginationButton>
      </PaginationItem>
    </PaginationList>
  )
}

Pagination.defaultProps = {
  totalPages: null,
  currentPage: 1,
}

Pagination.propTypes = {
  currentPage: PropTypes.number,
  totalPages: PropTypes.number,
  filter: PropTypes.string.isRequired,
  fetchProducts: PropTypes.func.isRequired,
}

export default Pagination
