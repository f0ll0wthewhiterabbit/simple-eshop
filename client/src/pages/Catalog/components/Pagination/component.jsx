import React from 'react'
import PropTypes from 'prop-types'
import ChevronRightIcon from '@material-ui/icons/ChevronRight'
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft'

import { Wrapper, PaginationList, PaginationItem, PaginationButton, Label } from './styles'

const Pagination = ({
  currentPage,
  totalPages,
  totalAmount,
  filter,
  withLabel,
  withNumbers,
  fetchProductsRequest,
}) => {
  const pageNumbers = Array.from({ length: totalPages }, (v, k) => k + 1)

  const handlePageNumberClick = event => {
    fetchProductsRequest(Number(event.target.innerText), { filter })
  }

  const handlePreviousPageClick = () => {
    fetchProductsRequest(currentPage - 1, { filter })
  }
  const handleNextPageClick = () => {
    fetchProductsRequest(currentPage + 1, { filter })
  }

  const isPageNumberInPagination = pageNumber => {
    if (!withNumbers) {
      return pageNumber === currentPage
    }

    return (
      pageNumber === 1 ||
      pageNumber === totalPages ||
      (pageNumber >= currentPage - 2 && pageNumber <= currentPage + 2)
    )
  }

  return (
    <Wrapper>
      {withLabel && <Label>{totalAmount} Products</Label>}
      <PaginationList>
        <PaginationItem>
          <PaginationButton
            aria-label="previous"
            size="small"
            disabled={currentPage === 1 ? true : undefined}
            onClick={handlePreviousPageClick}
            data-test="previousButton"
          >
            <ChevronLeftIcon />
          </PaginationButton>
        </PaginationItem>
        {pageNumbers.map(
          pageNumber =>
            isPageNumberInPagination(pageNumber) && (
              <PaginationItem key={pageNumber} data-test="paginationItem">
                <PaginationButton
                  type="button"
                  active={pageNumber === currentPage ? 1 : undefined}
                  onClick={pageNumber === currentPage ? undefined : handlePageNumberClick}
                  data-test={`paginationButton${pageNumber}`}
                >
                  {pageNumber}
                </PaginationButton>
              </PaginationItem>
            )
        )}
        <PaginationItem>
          <PaginationButton
            aria-label="next"
            size="small"
            disabled={currentPage === totalPages ? true : undefined}
            onClick={handleNextPageClick}
            data-test="nextButton"
          >
            <ChevronRightIcon />
          </PaginationButton>
        </PaginationItem>
      </PaginationList>
    </Wrapper>
  )
}

Pagination.defaultProps = {
  totalPages: null,
  currentPage: 1,
  withLabel: false,
  withNumbers: true,
}

Pagination.propTypes = {
  currentPage: PropTypes.number,
  totalPages: PropTypes.number,
  totalAmount: PropTypes.number.isRequired,
  filter: PropTypes.string.isRequired,
  withLabel: PropTypes.bool,
  withNumbers: PropTypes.bool,
  fetchProductsRequest: PropTypes.func.isRequired,
}

export default Pagination
