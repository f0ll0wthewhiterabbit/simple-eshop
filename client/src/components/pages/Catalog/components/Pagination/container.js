import { connect } from 'react-redux'

import Pagination from './component'
import { fetchProducts } from '../../../../../store/actions'

const mapStateToProps = state => ({
  currentPage: state.getIn(['products', 'currentPage']),
  totalPages: state.getIn(['products', 'totalPages']),
  filter: state.getIn(['products', 'filter']),
})

const mapDispatchToProps = dispatch => ({
  fetchProducts: (currentPage, itemsPerPage, filter) =>
    dispatch(fetchProducts(currentPage, itemsPerPage, filter)),
})

export default connect(mapStateToProps, mapDispatchToProps)(Pagination)
