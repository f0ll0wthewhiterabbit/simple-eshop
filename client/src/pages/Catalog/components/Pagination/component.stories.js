import React from 'react'
import { action } from '@storybook/addon-actions'
import { withKnobs, object, number } from '@storybook/addon-knobs/react'

import Pagination from './component'

export default {
  title: 'Pagination',
  component: Pagination,
  excludeStories: /.*Data$/,
  decorators: [withKnobs],
}

export const paginationData = {
  currentPage: 5,
  totalPages: 10,
  totalAmount: 85,
  withLabel: false,
  withNumbers: true,
}

export const actionsData = {
  fetchProducts: action('fetchProducts'),
  confirmMethod: action('confirmMethod'),
}

export const normal = () => (
  <Pagination {...object('pagination', { ...paginationData })} {...actionsData} />
)
export const withLabel = () => (
  <Pagination
    {...paginationData}
    {...actionsData}
    withLabel
    totalAmount={number('totalAmount', 85, {
      range: true,
      min: 80,
      max: 90,
      step: 1,
    })}
  />
)
export const withCurrentPageNumberOnly = () => (
  <Pagination {...paginationData} {...actionsData} withNumbers={false} />
)
export const firstPage = () => <Pagination {...paginationData} currentPage={1} {...actionsData} />
export const lastPage = () => (
  <Pagination {...paginationData} currentPage={paginationData.totalPages} {...actionsData} />
)
