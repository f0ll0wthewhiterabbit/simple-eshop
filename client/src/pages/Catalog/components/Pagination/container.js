import { connect } from 'react-redux'

import Pagination from './component'
import { fetchProducts } from '../../../../store/actions'

const mapStateToProps = state => ({
  currentPage: state.getIn(['products', 'currentPage']),
  totalPages: state.getIn(['products', 'totalPages']),
  totalAmount: state.getIn(['products', 'totalAmount']),
  filter: state.getIn(['products', 'filter']),
})

const mapDispatchToProps = {
  fetchProducts: (page, itemsPerPage, filter) => fetchProducts(page, itemsPerPage, filter),
}

export default connect(mapStateToProps, mapDispatchToProps)(Pagination)
