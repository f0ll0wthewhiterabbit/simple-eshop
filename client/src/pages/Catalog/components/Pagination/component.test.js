import React from 'react'
import { shallow } from 'enzyme'

import Pagination from './component'
import findByTestAttr from '../../../../utils/findByTestAttr'
import { FIELDS } from '../../../../constants'

describe('Pagination component', () => {
  let wrapper
  const filter = FIELDS.URL_RATINGS_FILTER
  const initialProps = {
    currentPage: 5,
    totalPages: 10,
    totalAmount: 85,
    filter: FIELDS.URL_RATINGS_FILTER,
    fetchProducts: jest.fn(),
  }
  const dataTestPaginationItem = 'paginationItem'
  const dataTestPerviousButton = 'previousButton'
  const dataTestNextButton = 'nextButton'

  const generateWrapper = passedProps => {
    const defaultProps = { ...initialProps }
    const props = { ...defaultProps, ...passedProps }

    return shallow(<Pagination {...props} />)
  }

  beforeEach(() => {
    wrapper = generateWrapper()
  })

  afterEach(() => {
    initialProps.fetchProducts.mockClear()
  })

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should render correct number for current page', () => {
    const paginationButton = findByTestAttr(wrapper, `paginationButton${initialProps.currentPage}`)

    expect(paginationButton.text()).toBe(initialProps.currentPage.toString())
  })

  it('should render current page button as active button', () => {
    const paginationButton = findByTestAttr(wrapper, `paginationButton${initialProps.currentPage}`)

    expect(paginationButton.props().active).toBeTruthy()
  })

  it('should contain correct page numbers', () => {
    const paginationItems = findByTestAttr(wrapper, dataTestPaginationItem)
    const expectedAmountOfButtons = 7

    expect(paginationItems).toHaveLength(expectedAmountOfButtons)
  })

  it('should contain correct page numbers for first page', () => {
    wrapper = generateWrapper({ currentPage: 1 })
    const paginationItems = findByTestAttr(wrapper, dataTestPaginationItem)
    const expectedAmountOfButtons = 4

    expect(paginationItems).toHaveLength(expectedAmountOfButtons)
  })

  it('should contain correct page numbers for last page', () => {
    wrapper = generateWrapper({ currentPage: initialProps.totalPages })
    const paginationItems = findByTestAttr(wrapper, dataTestPaginationItem)
    const expectedAmountOfButtons = 4

    expect(paginationItems).toHaveLength(expectedAmountOfButtons)
  })

  it('should contain disabled previous button for first page', () => {
    wrapper = generateWrapper({ currentPage: 1 })
    const previousButton = findByTestAttr(wrapper, dataTestPerviousButton)

    expect(previousButton.props().disabled).toBeTruthy()
  })

  it('should contain disabled previous button for last page', () => {
    wrapper = generateWrapper({ currentPage: initialProps.totalPages })
    const nextButton = findByTestAttr(wrapper, dataTestNextButton)

    expect(nextButton.props().disabled).toBeTruthy()
  })

  it('should handle page number click', () => {
    const pageNumber = 7
    const paginationButton = findByTestAttr(wrapper, `paginationButton${pageNumber}`)
    paginationButton.simulate('click', { target: { innerText: pageNumber.toString() } })

    expect(initialProps.fetchProducts).toHaveBeenLastCalledWith(pageNumber, null, undefined, filter)
  })

  it('should handle previous page click', () => {
    const previousButton = findByTestAttr(wrapper, dataTestPerviousButton)
    previousButton.simulate('click')

    expect(initialProps.fetchProducts).toHaveBeenLastCalledWith(
      initialProps.currentPage - 1,
      null,
      undefined,
      filter
    )
  })

  it('should handle next page click', () => {
    const nextButton = findByTestAttr(wrapper, dataTestNextButton)
    nextButton.simulate('click')

    expect(initialProps.fetchProducts).toHaveBeenLastCalledWith(
      initialProps.currentPage + 1,
      null,
      undefined,
      filter
    )
  })

  it('should handle active page button click', () => {
    const paginationButton = findByTestAttr(wrapper, `paginationButton${initialProps.currentPage}`)
    paginationButton.simulate('click')

    expect(initialProps.fetchProducts).toHaveBeenCalledTimes(0)
  })
})
