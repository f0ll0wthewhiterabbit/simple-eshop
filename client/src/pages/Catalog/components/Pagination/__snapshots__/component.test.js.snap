// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component should render correctly 1`] = `
<styled.div>
  <styled.ul>
    <styled.li>
      <Styled(WithStyles(ForwardRef(IconButton)))
        aria-label="previous"
        data-test="previousButton"
        onClick={[Function]}
        size="small"
      >
        <ChevronLeftIcon />
      </Styled(WithStyles(ForwardRef(IconButton)))>
    </styled.li>
    <styled.li
      data-test="paginationItem"
      key="1"
    >
      <Styled(WithStyles(ForwardRef(IconButton)))
        data-test="paginationButton1"
        onClick={[Function]}
        type="button"
      >
        1
      </Styled(WithStyles(ForwardRef(IconButton)))>
    </styled.li>
    <styled.li
      data-test="paginationItem"
      key="3"
    >
      <Styled(WithStyles(ForwardRef(IconButton)))
        data-test="paginationButton3"
        onClick={[Function]}
        type="button"
      >
        3
      </Styled(WithStyles(ForwardRef(IconButton)))>
    </styled.li>
    <styled.li
      data-test="paginationItem"
      key="4"
    >
      <Styled(WithStyles(ForwardRef(IconButton)))
        data-test="paginationButton4"
        onClick={[Function]}
        type="button"
      >
        4
      </Styled(WithStyles(ForwardRef(IconButton)))>
    </styled.li>
    <styled.li
      data-test="paginationItem"
      key="5"
    >
      <Styled(WithStyles(ForwardRef(IconButton)))
        active={1}
        data-test="paginationButton5"
        type="button"
      >
        5
      </Styled(WithStyles(ForwardRef(IconButton)))>
    </styled.li>
    <styled.li
      data-test="paginationItem"
      key="6"
    >
      <Styled(WithStyles(ForwardRef(IconButton)))
        data-test="paginationButton6"
        onClick={[Function]}
        type="button"
      >
        6
      </Styled(WithStyles(ForwardRef(IconButton)))>
    </styled.li>
    <styled.li
      data-test="paginationItem"
      key="7"
    >
      <Styled(WithStyles(ForwardRef(IconButton)))
        data-test="paginationButton7"
        onClick={[Function]}
        type="button"
      >
        7
      </Styled(WithStyles(ForwardRef(IconButton)))>
    </styled.li>
    <styled.li
      data-test="paginationItem"
      key="10"
    >
      <Styled(WithStyles(ForwardRef(IconButton)))
        data-test="paginationButton10"
        onClick={[Function]}
        type="button"
      >
        10
      </Styled(WithStyles(ForwardRef(IconButton)))>
    </styled.li>
    <styled.li>
      <Styled(WithStyles(ForwardRef(IconButton)))
        aria-label="next"
        data-test="nextButton"
        onClick={[Function]}
        size="small"
      >
        <ChevronRightIcon />
      </Styled(WithStyles(ForwardRef(IconButton)))>
    </styled.li>
  </styled.ul>
</styled.div>
`;
